### PR TITLE
Initial simple purification of method bodies 

### PR DESCRIPTION
**Pull request by tillarnold**

### CODE REVIEW

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 1]
```diff
@@ -470,15 +471,16 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
             .register(self.mir.span, ErrorCtxt::Unexpected);
         let method_with_fold_unfold = foldunfold::add_fold_unfold(
             self.encoder,
-            self.cfg_method,
+            self.cfg_method.clone(),
```
Comment (fpoli) [inline prusti-viper/src/encoder/procedure_encoder.rs]: Is the clone needed just to read the name of the method later? If so, you can store the method's name in a local variable before this call instead of cloning.

File: prusti-viper/src/encoder/encoder.rs [hunk 2]
```diff
@@ -89,7 +89,7 @@ pub struct Encoder<'v, 'tcx: 'v> {
     type_cast_functions: RefCell<HashMap<(ty::Ty<'tcx>, ty::Ty<'tcx>), vir::Function>>,
     memory_eq_encoder: RefCell<MemoryEqEncoder>,
     fields: RefCell<HashMap<String, vir::Field>>,
-    snapshots: RefCell<HashMap<String, Box<Snapshot>>>, // maps predicate names to snapshots
+    pub snapshots: RefCell<HashMap<String, Box<Snapshot>>>, // maps predicate names to snapshots
```
Comment (fpoli) [inline prusti-viper/src/encoder/encoder.rs]: Why is this needed?

File: prusti-viper/src/encoder/procedure_encoder.rs [hunk 3]
```diff
@@ -470,15 +471,16 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
...
             &loan_locations,
             &self.cfg_blocks_map,
             method_pos,
         )
         .map_err(|foldunfold_error| {
             SpannedEncodingError::internal(
                 format!(
-                    "generating fold-unfold Viper statements failed ({:?})",
+                    "generating fold-unfold Viper statements for method {:?} failed ({:?})",
+                    self.cfg_method.name(),
```
Comment (fpoli) [inline prusti-viper/src/encoder/procedure_encoder.rs]: `cfg_method.name()` is the Viper method name, so it's probably unreadable for the user. What about displaying the Rust's method name instead?

File: prusti-viper/src/encoder/encoder.rs (hunk 2 above)
Comment (vakaras) [reply prusti-viper/src/encoder/encoder.rs]: Please use the getters: https://github.com/viperproject/prusti-dev/blob/d0b6d9f8a56111f7c4269799e65b89bff5f6e16a/prusti-viper/src/encoder/encoder.rs#L1039-L1049

**Comment by tillarnold** [comment]
@vakaras I think I have cleaned this up to a somewhat acceptable degree

**Comment by vakaras** [comment]
> @vakaras I think I have cleaned this up to a somewhat acceptable degree

Thanks!
